Lib/xml/etree/ElementTree.py

```diff
@@ -508,6 +511,58 @@ def __eq__(self, other):
 
 # --------------------------------------------------------------------
 
+class XMLDeclarationQuotes(enum.Enum):
+    """
+    Whether or not single quotes or double quotes ought to be used in the XML
+    declaration.
+
+    *SINGLE* (default): <?xml version='1.0' encoding='UTF-8'?>
+    *DOUBLE*: <?xml version="1.0" encoding="UTF-8"?>
+    """
+    SINGLE = "'"
+    DOUBLE = '"'
+
+    def __str__(self):
+        return self.value
+
+class ShortEmptyElements(enum.Enum):
+    """
+    This class creates backwards compatibility with the boolean value of
+    *short_empty_elements* that existed prior to 3.??.
+
+    Assuming the tag `<q/>`, the results will be:
+
+    *SPACE* (default): `<q />`
+    *NOSPACE*: `<q/>`
+    *NONE*: `<q></q>`
+    """
+    SPACE = " "
+    NOSPACE = ""
+    NONE = False
+
+    def __bool__(self):
+        return self != ShortEmptyElements.NONE
+
+    @classmethod
+    def _missing_(cls, value):
+        if value is enum.no_arg:
+            return cls.SPACE
+        elif isinstance(value, bool):
+            return cls.SPACE if value else cls.NONE
+        else:
+            return super()._missing_(value)
+
+    @classmethod
+    def tag_defaultdict(cls, short_empty_elements):
+        if not isinstance(short_empty_elements, collections.defaultdict):
+            if isinstance(short_empty_elements, ShortEmptyElements):
+                return collections.defaultdict(lambda: short_empty_elements)
```

Comment:
I would also allow mappings that implement `__missing__`. Therefore, I suggest checking for everything else and assuming that what remains is probably some kind of acceptable dict. Or check for `collections.abc.Mapping` instead.

Reply:
Sorry I don't quite understand. Should I just check for `hasattr('__missing__')` and then try to call it and make sure its result is a `ShortEmptyElements`?

Reply:
I suggested to accept any Mapping, basically. You can either check `isinstance(..., collections.abc.Mapping)`, or you could also ducktypingly check for the enums first and simply assume that, otherwise, users probably passed some kind of acceptable mapping, and will have to like with a runtime error if the value misbehaves somehow.

There are really different ways for users to make an item lookup (subscripting) return a reasonable value that we can use for our processing here. A `defaultdict` is just one in many.